PolyEngine/Core/Src/Queue.hpp

```diff
@@ -1,25 +1,226 @@
+		T& Front() const { HEAVY_ASSERTE(!IsEmpty(), "Trying to access empty queue!"); return Data[Head]; };
+		T& Back() const { HEAVY_ASSERTE(!IsEmpty(), "Trying to access empty queue!"); return Data[GetPrevIdx(Tail)]; };
+
+		//------------------------------------------------------------------------------
+		void Reserve(size_t capacity)
+		{
+			HEAVY_ASSERTE(capacity >= Capacity, "Capacity too small!");
+			Realloc(capacity);
+		}
+
+		//------------------------------------------------------------------------------
+		bool Contains(const T& rhs) const { return FindIdx(rhs) < GetSize(); }
+
 	private:
-		Dynarray<T> Data;
+		//------------------------------------------------------------------------------
+		size_t GetPrevIdx(size_t idx) const { return (idx == 0) ? Capacity - 1 : idx - 1; }
+		size_t GetNextIdx(size_t idx) const { return (idx + 1) % Capacity; }
+		void DecreaseIdx(size_t& idx) const { idx = GetPrevIdx(idx); }
+		void AdvanceIdx(size_t& idx) const { idx = GetNextIdx(idx); }
+		size_t GetNthIdx(size_t& n)	const { return (Head + n) % Capacity; }
+
+		//------------------------------------------------------------------------------
+		size_t FindIdx(const T& rhs) const
+		{
+			for (size_t idx = Head; idx != Tail; AdvanceIdx(idx))
+				if (Data[idx] == rhs)
+					return idx;
+
+			return GetSize();
+		}
+
+		//------------------------------------------------------------------------------
+		void Realloc(size_t capacity)
+		{
+			HEAVY_ASSERTE(Size <= capacity, "Invalid resize capacity!");
+			T* newData = static_cast<T*>(DefaultAlloc(capacity * sizeof(T)));
+
+			// move all elements
+			for (size_t i = 0; i < Size; ++i)
+			{
+				size_t oldIdx = GetNthIdx(i);
+				ObjectLifetimeHelper::MoveCreate(newData + i, std::move(Data[oldIdx]));
+				ObjectLifetimeHelper::Destroy(Data + oldIdx);
+			}
+
+			DefaultFree(Data);
+			Data = newData;
+			Capacity = capacity;
+			Head = 0;
+			Tail = Size;
+		}
+
+		void Free() { if (Data) DefaultFree(Data); }
+
+		//------------------------------------------------------------------------------
+		void Copy(const Queue<T>& rhs)
+		{
+			Reserve(rhs.GetSize() + 1);
+			for (size_t idx = rhs.Head; idx != rhs.Tail; rhs.AdvanceIdx(idx))
+				PushBack(rhs.Data[idx]);
+		}
+
+		//------------------------------------------------------------------------------
+		void Move(Queue<T>&& rhs)
+		{
+			Size = rhs.Size;
+			Capacity = rhs.Capacity;
+			Data = rhs.Data;
+			Head = rhs.Head;
+			Tail = rhs.Tail;
+			rhs.Data = nullptr;
+			rhs.Size = 0;
+			rhs.Capacity = 0;
+			rhs.Head = 0;
+			rhs.Tail = 0;
+		}
+
+		//------------------------------------------------------------------------------
+		void PopulateFromInitializerList(const std::initializer_list<T>& list)
+		{
+			Reserve(list.size());
+			for (const T& obj : list)
+				PushBack(obj);
+		}
+
+		void Enlarge() { Reserve((Capacity + 1) * 2); }
+
+		size_t Head = 0;
+		size_t Tail = 0;
+		size_t Size = 0;
+		size_t Capacity = 1;
```

Comment:
Capacity should be 0 on start. No need to allocate memory if queue won't be used (for some reason). This is working better when someone is calling Reserve just after calling empty constructor.